File: split_log.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
按线程ID分割日志文件。
默认处理 output 目录中最新的 img2text_*.log，也可通过 --logfile 指定。
输出：在日志文件所在目录创建同名文件夹（去掉 .log），每个线程生成
      Txx_<日志名>.log 文件（T00 为主线程/其它未命名线程）。
"""

import re
import sys
import argparse
from pathlib import Path
from collections import defaultdict

# 匹配日志行开头的线程ID，例如 [17:43:06][T01]
THREAD_PAT = re.compile(r'^\[.*?\]\[(T\d+)\]')


def find_latest_log(output_dir: Path) -> Path:
    """在指定目录查找最新的 img2text_*.log"""
    logs = sorted(output_dir.glob("img2text_*.log"), key=lambda p: p.stat().st_mtime, reverse=True)
    if not logs:
        raise FileNotFoundError(f"No img2text_*.log found in {output_dir}")
    return logs[0]


def split_log(log_path: Path):
    # 读取全部行
    with open(log_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    # 按线程分组，同时保持原始顺序（组内追加）
    thread_lines = defaultdict(list)

    for line in lines:
        m = THREAD_PAT.match(line)
        tid = m.group(1) if m else "T00"  # 没有线程ID的行归入 T00
        thread_lines[tid].append(line)

    # 准备输出文件夹：与日志文件同名（去掉 .log 后缀）
    base_name = log_path.stem  # 例如 img2text_20260429_174305
    out_dir = log_path.parent / base_name
    out_dir.mkdir(exist_ok=True)

    # 为每个线程写出文件
    for tid in sorted(thread_lines.keys()):
        out_file = out_dir / f"{tid}_{base_name}.log"
        with open(out_file, 'w', encoding='utf-8') as f:
            f.writelines(thread_lines[tid])
        print(f"  Wrote {len(thread_lines[tid])} lines to {out_file.name}")

    print(f"\nSplit into {len(thread_lines)} thread files in {out_dir}/")


def main():
    parser = argparse.ArgumentParser(description="Split log by thread ID")
    parser.add_argument("--logfile", type=str, help="Path to log file")
    parser.add_argument("--output-dir", type=str, default=None,
                        help="Directory containing logs (default: from config.yaml or current dir)")
    parser.add_argument("--config", type=str, default="config.yaml",
                        help="Config file to read output_dir (used when --logfile not given)")
    args = parser.parse_args()

    if args.logfile:
        log_path = Path(args.logfile)
    else:
        # 尝试从 config 读取 output_dir
        out_dir = Path.cwd() / "finally"   # 默认 fallback
        if Path(args.config).exists():
            import yaml
            try:
                with open(args.config, 'r', encoding='utf-8') as f:
                    config = yaml.safe_load(f)
                out_dir = Path(config["paths"]["output_dir"])
            except Exception:
                pass  # 使用默认

        log_path = find_latest_log(out_dir)
        print(f"Using latest log: {log_path.name}")

    if not log_path.exists():
        print(f"Error: log file not found: {log_path}")
        sys.exit(1)

    print(f"Splitting {log_path}")
    split_log(log_path)
    print("Done.")


if __name__ == "__main__":
    main()

File: go/internal/splitlog/splitlog.go
// Package splitlog splits an img2text_*.log file into per-thread files
// by parsing the leading "[HH:MM:SS][Txx]" prefix on each line. Lines that
// do not match the prefix are grouped into "T00". Output files are written
// to a directory named after the input log (without the .log extension).
package splitlog

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"mineru-tools/internal/config"
)

// threadPat matches the leading "[HH:MM:SS][Txx]" prefix on a log line.
var threadPat = regexp.MustCompile(`^\[.*?\]\[(T\d+)\]`)

// FindLatestLog returns the path of the newest img2text_*.log file in
// logDir. Files are sorted by name in descending order, so the lexicographically
// largest filename (which embeds a timestamp) is returned.
func FindLatestLog(logDir string) (string, error) {
	entries, err := os.ReadDir(logDir)
	if err != nil {
		return "", fmt.Errorf("read log dir %s: %w", logDir, err)
	}

	var names []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		if strings.HasPrefix(name, "img2text_") && strings.HasSuffix(name, ".log") {
			names = append(names, name)
		}
	}
	if len(names) == 0 {
		return "", fmt.Errorf("no img2text_*.log found in %s", logDir)
	}

	sort.Sort(sort.Reverse(sort.StringSlice(names)))
	return filepath.Join(logDir, names[0]), nil
}

// SplitLogByThread reads the log file at logPath, groups lines by their
// leading thread ID, and writes one file per thread. If outputDir is
// non-empty, files are written there; otherwise a sibling directory named
// after the log (without the .log extension) is used.
func SplitLogByThread(logPath, outputDir string) error {
	f, err := os.Open(logPath)
	if err != nil {
		return fmt.Errorf("open log %s: %w", logPath, err)
	}
	defer f.Close()

	threadLines := make(map[string][]string)
	scanner := bufio.NewScanner(f)
	// Allow long log lines.
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := scanner.Text()
		m := threadPat.FindStringSubmatch(line)
		tid := "T00"
		if m != nil {
			tid = m[1]
		}
		threadLines[tid] = append(threadLines[tid], line)
	}
	if err := scanner.Err(); err != nil {
		return fmt.Errorf("read log %s: %w", logPath, err)
	}

	// Output directory: explicit override, else default to a sibling
	// directory named after the log's stem.
	if outputDir == "" {
		dir := filepath.Dir(logPath)
		base := strings.TrimSuffix(filepath.Base(logPath), ".log")
		outputDir = filepath.Join(dir, base)
	}
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return fmt.Errorf("create out dir %s: %w", outputDir, err)
	}

	tids := make([]string, 0, len(threadLines))
	for tid := range threadLines {
		tids = append(tids, tid)
	}
	sort.Strings(tids)

	base := strings.TrimSuffix(filepath.Base(logPath), ".log")
	for _, tid := range tids {
		outFile := filepath.Join(outputDir, fmt.Sprintf("%s_%s.log", tid, base))
		if err := writeLines(outFile, threadLines[tid]); err != nil {
			return err
		}
		fmt.Printf("  Wrote %d lines to %s\n", len(threadLines[tid]), filepath.Base(outFile))
	}

	fmt.Printf("\nSplit into %d thread files in %s/\n", len(threadLines), outputDir)
	return nil
}

// Run is the entry point used by the CLI: if logFile is empty, the latest
// img2text_*.log in cfg.Paths.FinallyDir is used. If outputDir is non-empty,
// it overrides the default sibling-directory placement. It prints a summary
// of the created files via SplitLogByThread.
func Run(cfg *config.Config, logFile, outputDir string) error {
	if logFile == "" {
		if cfg == nil {
			return fmt.Errorf("config and logFile are both empty")
		}
		latest, err := FindLatestLog(cfg.Paths.FinallyDir)
		if err != nil {
			return err
		}
		logFile = latest
		fmt.Printf("Using latest log: %s\n", filepath.Base(logFile))
	}

	if _, err := os.Stat(logFile); err != nil {
		return fmt.Errorf("log file not found: %s", logFile)
	}

	fmt.Printf("Splitting %s\n", logFile)
	return SplitLogByThread(logFile, outputDir)
}

func writeLines(path string, lines []string) error {
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("create %s: %w", path, err)
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for _, l := range lines {
		if _, err := w.WriteString(l); err != nil {
			return fmt.Errorf("write %s: %w", path, err)
		}
		if !strings.HasSuffix(l, "\n") {
			if _, err := w.WriteString("\n"); err != nil {
				return fmt.Errorf("write %s: %w", path, err)
			}
		}
	}
	return w.Flush()
}
